cpp/src/arrow/array/builder_adaptive.cc

```diff
@@ -124,7 +135,14 @@ Status AdaptiveIntBuilder::AppendValuesInternal(const int64_t* values, int64_t l
     const int64_t chunk_size = std::min(length, kAdaptiveIntChunkSize);
 
     uint8_t new_int_size;
-    new_int_size = internal::DetectIntWidth(values, valid_bytes, chunk_size, int_size_);
+    if (values == reinterpret_cast<const int64_t*>(pending_data_)) {
```

Comment:
This should be reverted now that type() is virtual.

Reply:
(allows the removal of the branch https://github.com/apache/arrow/pull/4930/files#diff-6a872d5322a18ef2f9bd1d5abbd75ba3R102 in Append)